product_variant_configurator/models/product_product.py

```diff
@@ -0,0 +1,106 @@
+        if description:
+            return "%s\n%s" % (
+                self.product_tmpl_id.name, "\n".join(description))
+        else:
+            return self.product_tmpl_id.name
+
+    @api.model
+    def _build_attributes_domain(self, product_template, product_attributes):
+        domain = []
+        cont = 0
+        if product_template:
+            domain.append(('product_tmpl_id', '=', product_template.id))
+            for attr_line in product_attributes:
+                if isinstance(attr_line, dict):
+                    value_id = attr_line.get('value_id')
+                else:
+                    value_id = attr_line.value_id.id
+                if value_id:
+                    domain.append(('attribute_value_ids', '=', value_id))
+                    cont += 1
+        return domain, cont
+
+    @api.model
+    def _product_find(self, product_template, product_attributes):
+        if product_template:
+            domain, cont = self._build_attributes_domain(
+                product_template, product_attributes)
+            products = self.search(domain)
+            # Filter the product with the exact number of attributes values
+            for product in products:
+                if len(product.attribute_value_ids) == cont:
+                    return product
+        return False
+
+    @api.constrains('product_tmpl_id', 'attribute_value_ids')
+    def _check_duplicity(self):
+        for product in self:
+            domain = [('product_tmpl_id', '=', product.product_tmpl_id.id)]
+            for value in product.attribute_value_ids:
+                domain.append(('attribute_value_ids', '=', value.id))
+            other_products = self.with_context(
+                active_test=False).search(domain)
+            # Filter the product with the exact number of attributes values
+            cont = len(product.attribute_value_ids)
+            for other_product in other_products:
+                if (len(other_product.attribute_value_ids) == cont and
+                        other_product != product):
+                    raise exceptions.ValidationError(
+                        _("There's another product with the same attributes."))
+
+    @api.constrains('product_tmpl_id', 'attribute_value_ids')
+    def _check_configuration_validity(self):
+        """The method checks that the current selection values are correct.
+
+        As default, the validity means that all the attributes
+        values are set. This can be overridden to set another rules.
+
+        :raises: exceptions.ValidationError: If the check is not valid.
+        """
+        for product in self:
+            if bool(product.product_tmpl_id.attribute_line_ids.mapped(
+                    'attribute_id') -
+                    product.attribute_value_ids.mapped('attribute_id')):
+                raise exceptions.ValidationError(
+                    _("You have to fill all the attributes values."))
```

Comment:
@pedrobaeza Its this line here, I'm definitely in favour of keeping it, otherwise you probably get errors with your grid thing in sale order.  What I would do though is on the product variant view under attribute lines is add a required flag to attribute_value_ids, so the user gets a normal required field error.  It makes no sense to add an attribute line without attribute values.

Also I'd change this error message to something more helpful, like "Each attribute line must have values assigned for variant creation" - or something.

Reply:
@gdgellatly Doing some some tests on the issue I see that this error only raises if you add more than one attribute-value at once. Otherwise it behaves as expected (creates variants as you add attributes and values). So I think that this check should be done in other way: https://github.com/Tecnativa/product-variant/blob/aabddabe8699fc0ec59729ddab1b5d85681bc858/product_variant_configurator/models/product_product.py#L91

Reply:
@chienandalu I double checked that constraint, and couldn't reproduce on runbot.  I still think the constraint is correct and should be there.  However we still have the issue of the original created product.  I think what it needs is an active test.  If product.active and (rest of constraint) because otherwise the constraint is applying to deactivated products, which in the case of added attributes will make them fail this test.

Reply:
So that's what I said at the beginning... @chienandalu please add that.

Reply:
also here is the error here 
```
if bool(product.product_tmpl_id.attribute_line_ids.mapped('attribute_id')
```
should be in pseudo code
```
if bool(product.product_tmpl_id.attribute_line_ids.filtered('attribute_value_ids' is set).mapped('attribute_id')
or
if bool(product.product_tmpl_id.attribute_line_ids.mapped('attribute_value_ids').mapped('attribute_id')
```
see the difference, one checks that the line has any attribute value ids before adding attributes to the recordset

Reply:
@pedrobaeza I read you wrong sorry, but also note the other error here, that will need addressing too, the case as per @cubells report where the line has no attributes.

Reply:
I think the error is trickier than that. Debugging into it I'm seeing that on this issue cases the constraint is checked in the middle of the product's write(). As a result of this, not all the attributes are written in that moment and the check fails. Checking if product is active had no effect.

Reply:
@chienandalu I'm not sure I quite understand the constraint anymore - I thought the reason for it was that when you created a manual variant, you had to ensure completion of one value from each attribute line.  This seems to just say the attribute lines must have values, which technically they don't need to.

But I'm going to drop out of this conversation, I have surgery today and won't be back to things for a little while.

Reply:
Indeed, that's not the purpose, so @chienandalu please revert to previous condition and we will check it together.

@gdgellatly have a good surgery and I wish you a soon recover.

Reply:
@pedrobaeza Ok, I see. I dropped the commit. That being so I find it difficult to overlook the false positives for the reasons we were discussing before (https://github.com/OCA/product-variant/pull/65#discussion_r135505113) :confused:

Reply:
@gdgellatly Seen with @pedrobaeza and hopefully fixed in  https://github.com/OCA/product-variant/pull/65/commits/b99c148c1c3e8119ac17474dc6343af023302c85